packages/read-model/src/Product/Listed/ListedProductViewInterface.php

```diff
@@ -0,0 +1,12 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Shopsys\ReadModelBundle\Product\Listed;
+
+/**
+ * @experimental
+ */
+interface ListedProductViewInterface
```

Comment:
Please remove the concept of interface for views as there is no added value.

```php
$topProducts = $this->listedProductsFacade->getTop();
foreach($topProducts as $product) {
    $product-> // nothing!
}
```

According to the interface I cannot do anything with the object. If you change it to actual `ListedProductView`, I'll know what can I get from it.

I understand you wanted to introduce interfaces for view objects because once the view is extended, the factory returns still the correct interface. But it also still returns the correct class anyway... So there is no added value.

It doesn't make sense generally to create an interface for read model objects, as same as for domain objects. The interface makes sense when you have a changeable behavior, eg. getting data.

Reply:
The main idea of these interfaces is not to model changeable behavior but provide type information about objects instead. It's known as a Marker Interface pattern. 

> According to the interface I cannot do anything with the object

To me, it's not a problem here because objects are directly used in twig templates only (and tests, but I'm ok with it there).


And the added value to me is honesty about type while extendind. Consider `createFromProduct` method from `ListedProductViewFactory` class. This method should return View object and is now defined like:
```php
protected function createFromProduct(Product $product, ?ImageViewInterface $imageView, ProductActionViewInterface $productActionView): ListedProductViewInterface
```

In a way how it's implemented, this method is first to be overridden if I need another field in the view object.

Let's presume, I declare this method to return the `ListedProductView` object directly. Like this:
```php
protected function createFromProduct(...): ReadModelBundle\ListedProductView
```
In my project, I cannot override this method and declare a different return type
```php
namespace Shopsys\ShopBundle\Model\Product\ViewFactory;

// Invalid: Return type declaration must be compatible with Shopsys\ReadModelBundle\Product\ListedProductViewFactory
protected function createFromProduct(...): ShopBundle\ListedProductView 
```

So I need to return parent class and then I lie about the return type – if I have the brand in my extended `ListedProductView`, returned object here looks like there is no brand at all, so in my point of view, this solves nothing.

And I'm not a fan of a different type in annotation and in typehint.

So Interface still seems to me like the cleanest possible solution.

Reply:
> So I need to return parent class and then I lie about the return type

This is not true. `B extends A` means `B IS A`, so if you declare `return A` and you return `B`, you don't lie.

> And the added value to me is honesty about type while extendind.

I don't see the added value. With both - interface and parent class declaration - you are honest and you don't lie.

> The main idea of these interfaces is not to model changeable behavior but provide type information about objects instead.

Type information is also provided by class declaration.


If you have an interface
```php
protected function createFromProduct(...): ListedProductViewInterface
```
You know almost nothing about returned object, only that it is somehow marked. What is this information good for?

If you have an interface with class
```php
protected function createFromProduct(...): ListedProductView
```
You know at least something about returned object. Definitely more than with the interface

Reply:
After a long discussion with @vitek-rostislav, with a heavy heart, I removed the Interfaces. 

I still believe it's a good concept but I understand your arguments.

Reply:
Thanks!
The interface topic deserves deeper explanation... I'd like to meet with you to not just explain but also to show when your approach makes sense... if you agree :-)

Reply:
I just checked all references of `ViewInterface` are gone :heavy_check_mark: